mock: speed up mock.findExpectedCall by skipping diff if not necessary

mrahs opened the pull request:
## Summary
Speed up `mock.findExpectedCall` by skipping argument diffing if not necessary

## Changes
Modified the function to skip over used up expected calls.
Everything else remains the same from the user's perspective, including the associated special printed message for such cases.

## Motivation
I have test suites with many one-time expected calls. I have noticed that the suite keeps accumulating such expired calls and evaluates every one of them before reaching the few usable ones.
Ideally, such calls should be discarded to save space and cycles, but that's an optimization for another day.

## Related issues
Closes #983

mrahs commented:
@boyan-soubachov I wonder if you can take a look?

boyan-soubachov reviewed (changes requested):
Thank you for this contribution.

I agree with simplifying (and optimising) the logic in `findExpectedCall` but it also looks to me like you're changing the behaviour of the exported function (`args.Diff`).

This means that this change would not be backwards compatible and would have to wait for testify v2.0.0.

Would it make more sense to split this into 2 PRs, one for optimising the iteration logic in `findExpectedCall` and the other for reducing the return vars and simplifying the upstream call-chain?

mrahs committed:
Skip argument diff if not necessary (#983)

Expected calls that have used up are still checked for matches on every
call. When the mock object has accumulated many expected calls it
becomes increasingly more expensive to calculate the diff for every
single expected call.

This improves things a bit by skipping the diff calculation of expected
calls that have been used up.

All changes are internal and should not change any external behaviour.

mrahs commented:
@boyan-soubachov thank you for the review. I have updated the code to preserve visible behavior and a added a test case for expected error when a method has been used up to ensure consistent behavior before and after this change.
